firedrake/formmanipulation.py

```diff
@@ -127,6 +128,56 @@ def argument(self, o):
                          for j in numpy.ndindex(V_is[i].value_shape)]
         return self._arg_cache.setdefault(o, as_vector(args))
 
+    def cofunction(self, o):
+        from firedrake import Cofunction, FunctionSpace, MixedFunctionSpace
+        from pyop2 import MixedDat
+        from ufl.classes import ZeroBaseForm
+
+        V = o.function_space()
+
+        # Not on a mixed space, just return ourselves.
+        if len(V) == 1:
+            return o
+
+        # We only need the test space for Cofunction
+        indices = self.blocks[0]
+        V_is = V.subfunctions
+
+        try:
+            indices = tuple(indices)
+            nidx = len(indices)
+        except TypeError:  # Only one index
+            indices = (indices, )
+            nidx = 1
+
+        # the cofunction should only be returned on the
+        # diagonal elements, so if we are off-diagonal
+        # on a two-form then we return a zero form,
+        # analogously to the off components of arguments.
+        if len(self.blocks) == 2:
+            itest, itrial = self.blocks
+            on_diag = (itest == itrial)
+        else:
+            on_diag = True
+
+        # if we are on the diagonal, then return a Cofunction
+        # in the relevant subspace that points to the data in
+        # the full space. This means that the right hand side
+        # of the fieldsplit problem will be correct.
+        if on_diag:
+            if nidx == 1:
+                i = indices[0]
+                W = V_is[i]
+                W = FunctionSpace(W.mesh(), W.ufl_element())  # primal space
+                c = Cofunction(W.dual(), val=o.subfunctions[i].dat)
+            else:
+                W = MixedFunctionSpace([V_is[i] for i in indices])  # dual space
```

Comment:
Isn't `V_is` a list of `Functions`? This doesn't look right to me.

The primal/dual split here also seems a bit confusing. Why should we have a primal thing if `len(indices) == 1` and a dual thing otherwise?

Reply:
`V_is` is a list of `FunctionSpaces`. I think the confusion is because when `Function.split` was renamed `Function.subfunctions`, the same change was made for spaces, i.e. `FunctionSpace.split` became `FunctionSpace.subfunctions`.

The primal/dual thing is annoying. If the split is for multiple fields then we need a `MixedFunctionSpace` and can just pass a list of dual spaces and it gives a mixed dual space. For the single field split case, `FunctionSpace` with the ufl_element gives a primal function space which is why the extra `W.dual()` is needed.
I'll see if there's a nicer way of doing it.

Reply:
Changed to using `DualSpace` to create the single component space. I'm sure I tried this before and got errors so I must have been doing something wrong then.